File: src/lib/toml/detect.ts
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";

export interface DetectionResult {
  runtime: string;
  build?: string;
  start?: string;
  confidence: "high" | "medium" | "low";
  reason: string;
}

// ---------- Detection rules ----------

function detectNode(dir: string): DetectionResult | null {
  const pkgPath = join(dir, "package.json");
  if (!existsSync(pkgPath)) return null;

  let pkg: {
    scripts?: Record<string, string>;
    main?: string;
    engines?: { node?: string };
  };
  try {
    pkg = JSON.parse(readFileSync(pkgPath, "utf-8"));
  } catch {
    return null;
  }

  // Determine Node.js version from various sources
  let runtime = "node22";
  const nodeVersion = resolveNodeVersion(dir, pkg.engines?.node);
  if (nodeVersion && nodeVersion >= 24) {
    runtime = "node24";
  }

  const build = pkg.scripts?.build ? "npm run build" : undefined;
  let start: string | undefined;
  if (pkg.scripts?.start) {
    start = "npm start";
  } else if (pkg.main) {
    start = `node ${pkg.main}`;
  }

  return {
    runtime,
    build,
    start,
    confidence: "high",
    reason: "Detected package.json",
  };
}

function resolveNodeVersion(dir: string, enginesNode?: string): number | null {
  // Check .node-version first (most specific)
  for (const file of [".node-version", ".nvmrc"]) {
    const filePath = join(dir, file);
    if (existsSync(filePath)) {
      const content = readFileSync(filePath, "utf-8").trim();
      const major = extractMajorVersion(content);
      if (major !== null) return major;
    }
  }

  // Fall back to engines.node from package.json
  if (enginesNode) {
    const major = extractMajorVersion(enginesNode);
    if (major !== null) return major;
  }

  return null;
}

function extractMajorVersion(version: string): number | null {
  // Handle "v22.1.0", "22.1.0", "22", ">=22", "^22", "~22", "lts/*", etc.
  const match = /(\d+)/.exec(version);
  if (match) {
    return Number(match[1]);
  }
  return null;
}

function detectGo(dir: string): DetectionResult | null {
  const goModPath = join(dir, "go.mod");
  if (!existsSync(goModPath)) return null;

  return {
    runtime: "go",
    build: "go build -o app .",
    start: "./app",
    confidence: "high",
    reason: "Detected go.mod",
  };
}

function detectPython(dir: string): DetectionResult | null {
  const hasRequirements = existsSync(join(dir, "requirements.txt"));
  const hasPyproject = existsSync(join(dir, "pyproject.toml"));
  if (!hasRequirements && !hasPyproject) return null;

  const reason = hasRequirements ? "Detected requirements.txt" : "Detected pyproject.toml";

  let start: string | undefined;

  // Try to detect the framework
  const framework = detectPythonFramework(dir);
  if (framework) {
    start = framework;
  }

  return {
    runtime: "python3.13",
    start,
    confidence: hasRequirements ? "high" : "medium",
    reason,
  };
}

function detectPythonFramework(dir: string): string | null {
  // Check requirements.txt for common frameworks
  const reqPath = join(dir, "requirements.txt");
  if (existsSync(reqPath)) {
    const content = readFileSync(reqPath, "utf-8").toLowerCase();
    if (content.includes("fastapi")) {
      return "uvicorn main:app --host 0.0.0.0";
    }
    if (content.includes("flask")) {
      return "flask run --host 0.0.0.0";
    }
    if (content.includes("django")) {
      return "python manage.py runserver 0.0.0.0:8000";
    }
  }

  // Check pyproject.toml for dependencies
  const pyprojectPath = join(dir, "pyproject.toml");
  if (existsSync(pyprojectPath)) {
    const content = readFileSync(pyprojectPath, "utf-8").toLowerCase();
    if (content.includes("fastapi")) {
      return "uvicorn main:app --host 0.0.0.0";
    }
    if (content.includes("flask")) {
      return "flask run --host 0.0.0.0";
    }
    if (content.includes("django")) {
      return "python manage.py runserver 0.0.0.0:8000";
    }
  }

  return null;
}

function detectRust(dir: string): DetectionResult | null {
  const cargoPath = join(dir, "Cargo.toml");
  if (!existsSync(cargoPath)) return null;

  let start: string | undefined;
  try {
    const content = readFileSync(cargoPath, "utf-8");
    // Try to extract [[bin]] name or package name
    const binMatch = /\[\[bin]]\s*\n\s*name\s*=\s*"([^"]+)"/.exec(content);
    if (binMatch) {
      start = `./target/release/${binMatch[1]}`;
    } else {
      const nameMatch = /\[package]\s*\n\s*name\s*=\s*"([^"]+)"/.exec(content);
      if (nameMatch) {
        start = `./target/release/${nameMatch[1]}`;
      }
    }
  } catch {
    // Fall through without a start command
  }

  return {
    runtime: "rust",
    build: "cargo build --release",
    start,
    confidence: "high",
    reason: "Detected Cargo.toml",
  };
}

function detectDocker(dir: string): DetectionResult | null {
  const dockerfilePath = join(dir, "Dockerfile");
  if (!existsSync(dockerfilePath)) return null;

  return {
    runtime: "from-image",
    confidence: "medium",
    reason: "Detected Dockerfile",
  };
}

// ---------- Public API ----------

const detectors = [detectNode, detectGo, detectPython, detectRust, detectDocker];

export function detectProject(dir: string): DetectionResult | null {
  for (const detect of detectors) {
    const result = detect(dir);
    if (result) return result;
  }
  return null;
}

File: hostd/internal/config/detect.go
package config

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// DetectionResult holds the auto-detected project configuration.
type DetectionResult struct {
	Runtime    string `json:"runtime"`
	Build      string `json:"build,omitempty"`
	Start      string `json:"start,omitempty"`
	Confidence string `json:"confidence"` // "high", "medium", "low"
	Reason     string `json:"reason"`
}

// DetectProject auto-detects the project runtime, build command, and start command.
func DetectProject(dir string) *DetectionResult {
	if r := detectNode(dir); r != nil {
		return r
	}
	if r := detectGo(dir); r != nil {
		return r
	}
	if r := detectPython(dir); r != nil {
		return r
	}
	if r := detectRust(dir); r != nil {
		return r
	}
	if r := detectDocker(dir); r != nil {
		return r
	}
	return nil
}

func detectNode(dir string) *DetectionResult {
	pkgPath := filepath.Join(dir, "package.json")
	data, err := os.ReadFile(pkgPath)
	if err != nil {
		return nil
	}

	var pkg struct {
		Scripts struct {
			Start string `json:"start"`
			Build string `json:"build"`
		} `json:"scripts"`
	}
	if err := json.Unmarshal(data, &pkg); err != nil {
		return nil
	}

	// Detect Node version from .node-version or .nvmrc
	runtime := "node22" // default
	for _, versionFile := range []string{".node-version", ".nvmrc"} {
		if data, err := os.ReadFile(filepath.Join(dir, versionFile)); err == nil {
			version := strings.TrimSpace(string(data))
			if strings.HasPrefix(version, "24") || strings.HasPrefix(version, "v24") {
				runtime = "node24"
			}
		}
	}

	start := pkg.Scripts.Start
	if start == "" {
		// Common patterns
		if fileExists(filepath.Join(dir, "server.js")) {
			start = "node server.js"
		} else if fileExists(filepath.Join(dir, "index.js")) {
			start = "node index.js"
		} else if fileExists(filepath.Join(dir, "app.js")) {
			start = "node app.js"
		}
	} else {
		start = "npm start"
	}

	build := ""
	if pkg.Scripts.Build != "" {
		build = "npm run build"
	}

	confidence := "high"
	if start == "" {
		confidence = "medium"
	}

	return &DetectionResult{
		Runtime:    runtime,
		Build:      build,
		Start:      start,
		Confidence: confidence,
		Reason:     "Detected package.json",
	}
}

func detectGo(dir string) *DetectionResult {
	if !fileExists(filepath.Join(dir, "go.mod")) {
		return nil
	}
	return &DetectionResult{
		Runtime:    "base",
		Build:      "go build -o app .",
		Start:      "./app",
		Confidence: "high",
		Reason:     "Detected go.mod",
	}
}

func detectPython(dir string) *DetectionResult {
	hasRequirements := fileExists(filepath.Join(dir, "requirements.txt"))
	hasPyproject := fileExists(filepath.Join(dir, "pyproject.toml"))
	if !hasRequirements && !hasPyproject {
		return nil
	}

	result := &DetectionResult{
		Runtime:    "python3.13",
		Confidence: "medium",
		Reason:     "Detected Python project",
	}

	if hasRequirements {
		result.Build = "pip install -r requirements.txt"
	} else {
		result.Build = "pip install ."
	}

	// Detect framework
	if fileExists(filepath.Join(dir, "manage.py")) {
		result.Start = "python manage.py runserver 0.0.0.0:8000"
		result.Confidence = "high"
		result.Reason = "Detected Django project"
	} else if containsInFile(filepath.Join(dir, "requirements.txt"), "fastapi") {
		result.Start = "uvicorn main:app --host 0.0.0.0 --port 8000"
		result.Reason = "Detected FastAPI project"
	} else if containsInFile(filepath.Join(dir, "requirements.txt"), "flask") {
		result.Start = "flask run --host 0.0.0.0 --port 8000"
		result.Reason = "Detected Flask project"
	}

	return result
}

func detectRust(dir string) *DetectionResult {
	cargoPath := filepath.Join(dir, "Cargo.toml")
	if !fileExists(cargoPath) {
		return nil
	}

	// Try to extract binary name from Cargo.toml
	start := "./app"
	if data, err := os.ReadFile(cargoPath); err == nil {
		content := string(data)
		// Simple extraction: look for name = "..."
		if idx := strings.Index(content, "name = \""); idx != -1 {
			rest := content[idx+8:]
			if end := strings.Index(rest, "\""); end != -1 {
				start = "./" + rest[:end]
			}
		}
	}

	return &DetectionResult{
		Runtime:    "base",
		Build:      "cargo build --release && cp target/release/* ./app 2>/dev/null || true",
		Start:      start,
		Confidence: "high",
		Reason:     "Detected Cargo.toml",
	}
}

func detectDocker(dir string) *DetectionResult {
	if !fileExists(filepath.Join(dir, "Dockerfile")) {
		return nil
	}
	return &DetectionResult{
		Runtime:    "base",
		Confidence: "low",
		Reason:     "Detected Dockerfile (manual configuration recommended)",
	}
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func containsInFile(path, substr string) bool {
	data, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	return strings.Contains(strings.ToLower(string(data)), strings.ToLower(substr))
}
